aptos-move/framework/aptos-framework/sources/AptosGovernance.move

```diff
@@ -0,0 +1,185 @@
+/**
+ * TODO
+ */
+module AptosFramework::AptosGovernance {
+    use Std::Errors;
+    use Std::Event::{Self, EventHandle};
+    use Std::Option;
+    use Std::Signer;
+
+    use AptosFramework::Stake;
+    use AptosFramework::SystemAddresses;
+    use AptosFramework::Timestamp;
+    use AptosFramework::Voting;
+
+    /// Error codes.
+    const EINSUFFICIENT_PROPOSER_STAKE: u64 = 1;
+    const ENOT_DELEGATED_VOTER: u64 = 2;
+    const EINSUFFICIENT_STAKE_LOCKUP: u64 = 3;
+
+    struct GovernanceCapability has store, drop {}
+
+    struct AptosProposal has store, drop {
+        governance_capability: GovernanceCapability,
+    }
+
+    struct GovernanceConfig has key {
+        min_voting_threshold: u64,
+        required_proposer_stake: u64,
+        voting_period_secs: u64,
+    }
+
+    struct GovernanceEvents has key {
+        create_proposal_events: EventHandle<CreateProposalEvent>,
+        update_config_events: EventHandle<UpdateConfigEvent>,
+        vote_events: EventHandle<VoteEvent>,
+    }
+
+    struct CreateProposalEvent has drop, store {
+        proposer: address,
+        stake_pool: address,
+        proposal_id: u64,
+        execution_hash: vector<u8>,
+    }
+
+    struct VoteEvent has drop, store {
+        proposal_id: u64,
+        voter: address,
+        stake_pool: address,
+        num_votes: u64,
+        should_pass: bool,
+    }
+
+    struct UpdateConfigEvent has drop, store {
+        min_voting_threshold: u64,
+        required_proposer_stake: u64,
+        voting_period_secs: u64,
+    }
+
+    public fun initialize(
+        core_framework: &signer,
+        min_voting_threshold: u64,
+        required_proposer_stake: u64,
+        voting_period_secs: u64,
+    ) {
+        SystemAddresses::assert_core_framework(core_framework);
+
+        Voting::register<AptosProposal>(core_framework);
+        move_to(core_framework, GovernanceConfig {
+            voting_period_secs,
+            min_voting_threshold,
+            required_proposer_stake,
+        });
+        move_to(core_framework, GovernanceEvents {
+            create_proposal_events: Event::new_event_handle<CreateProposalEvent>(core_framework),
+            update_config_events: Event::new_event_handle<UpdateConfigEvent>(core_framework),
+            vote_events: Event::new_event_handle<VoteEvent>(core_framework),
+        });
+    }
+
+    public fun update_governance_config(
+        _proposal: AptosProposal,
```

Comment:
this probably should be GovernanceCapability

Reply:
We can embed GovernanceCapability inside AptosProposal, but it's a bit redundant since on AptosGovernance can create and hand out AptosProposal anyway. This is the main point of this design because it could save the Voting module from being aware of a second type. If users of Voting need granular capabilities for different actions, they can embed them into the proposal that will be passed back at resolution time and then check for granular capabilities during resolution.